packages/squiggle-lang/src/utility/result.ts

```diff
@@ -78,3 +78,11 @@ export function getError<T, E>(r: result<T, E>): E | undefined {
     return undefined;
   }
 }
+
+export function getExt<T, E>(r: result<T, E>): T {
+  if (r.ok) {
+    return r.value;
+  } else {
+    throw r.value;
```

Comment:
In some cases, `E` type is not a JS `Error`; throwing anything other than Error objects in JS is possible but discouraged.
My best guess about what's going to happen in case of, e.g., `Dist.sum`, is:
- this function will throw `DistError` (which is an object, `{ type: "OperationError", ... }`)
- `rethrowWithFrameStack` in `IError.ts` will catch it:
```typescript
  } else if (err instanceof Error) {
    throw IRuntimeError.fromMessageWithFrameStack(
      new REJavaScriptExn(err.message, err.name),
      frameStack
    );
  } else {
    throw IRuntimeError.fromMessageWithFrameStack(
      new REOther("Unknown exception"),
      frameStack
    );
  }
```

And then it will rethrow it with "Unknown exception", instead of `distErrorToString`, since it doesn't know anything about the error and how to serialize it.

Options:
1) refactor `DistError` to be a class, similar to `ErrorMessage` class hierarchy (possible but annoying)
2) pass a `(err: E) => Error` callback to `getExt`, maybe?
3) Avoid this function and wrap the error correctly in `binaryOperations.ts`, as you said in your comment there.

Last option seems best to me, and tbh I'd just avoid `accumulate` in `binaryOperations.ts` and write it in imperative style, then it's easy to stop the processing early; something like this:

```typescript
const results: BaseDist[] = [];
for (let i = 0; i < dists.length; i++) {
  const result = i ? binaryOperations.algebraicAdd(results[i-1], dist[i], { env }) : dist[i];
  if (!result.ok) return result;
  results.push(result);
}
return Ok(results);
```

(I know it looks less elegant than `accumulate` version, but it's not that bad)

PS: Throwing `REDistributionError` in `binaryOperations.ts` is unfortunately not a (good) option, because it violates separation of concerns; `dist/*` modules shouldn't use reducer errors if we want to extract them to a separate library eventually.

Reply:
Happy to do (3).